compiler/src/intermediate_representation/constraint_bucket.rs

```diff
@@ -112,21 +114,61 @@ impl WriteLLVMIR for ConstraintBucket {
+                    Instruction::Call(b) => {
+                        for arg_ty in &b.argument_types {
+                            if arg_ty.size > 1 {
+                                todo!("not yet handling call arg array logic");
+                            }
+                            assert_ne!(0, arg_ty.size, "size should be non-zero");
+                        }
+                        1
+                    },
+                    _ => unreachable!("Instruction {:#?} should not be used for constraint substitution", i),
+                };
+                assert_ne!(0, size, "must have non-zero size");
+                if size == 1 {
+                    let lhs = get_instruction_arg(prev.into_instruction_value(), STORE_DST_IDX);
+                    assert_eq!(bigint_type(producer).as_any_type_enum(), lhs.get_type(), "wrong type");
+                    let rhs_ptr = get_instruction_arg(prev.into_instruction_value(), STORE_SRC_IDX);
+                    let rhs = create_load(producer, rhs_ptr.into_pointer_value());
+                    let constr = new_constraint(producer);
+                    let call = create_call(
+                        producer,
+                        CONSTRAINT_VALUES_FN_NAME,
+                        &[
+                            to_basic_metadata_enum(lhs),
+                            to_basic_metadata_enum(rhs),
+                            to_basic_metadata_enum(constr),
+                        ],
+                    );
+                    Some(call)
+                } else {
+                    let lhs_ptr = get_instruction_arg(prev.into_instruction_value(), STORE_DST_IDX).into_pointer_value();
+                    assert_eq!(bigint_type(producer).ptr_type(Default::default()), lhs_ptr.get_type(), "wrong type");
+                    let rhs_ptr = get_instruction_arg(prev.into_instruction_value(), STORE_SRC_IDX).into_pointer_value();
+
+                    let constraint_calls: Vec<_> = (0..size).map(|i| {
+                        let idx = create_literal_u32(producer, i as u64);
+                        let lhs = create_load(producer, create_gep(producer, lhs_ptr, &[idx]).into_pointer_value());
+                        let rhs = create_load(producer, create_gep(producer, rhs_ptr, &[idx]).into_pointer_value());
+                        let constr = new_constraint_with_name(producer, format!("constraint_{}", i).as_str());
+                        create_call(
+                            producer,
+                            CONSTRAINT_VALUES_FN_NAME,
+                            &[
+                                to_basic_metadata_enum(lhs),
+                                to_basic_metadata_enum(rhs),
+                                to_basic_metadata_enum(constr),
+                            ],
+                        )
+                    }).collect();
+
+                    Some(constraint_calls.last().expect("must have >1 value!").clone())
```

Comment:
I don't think we want a clone here. If that causes a borrow check issue, then you probably need to convert the map above to a normal loop like the BlockBucket implementation.

Reply:
Gotcha. I got overzealous with fun rust features.